classes/external/get_join_url.php

```
@@ -0,0 +1,121 @@
+        ]);
+    }
+
+    /**
+     * Updates a recording
+     *
+     * @param int $cmid the bigbluebuttonbn course module id
+     * @param null|int $groupid
+     * @return array (empty array for now)
+     * @throws \restricted_context_exception
+     */
+    public static function execute(
+        int $cmid,
+        ?int $groupid = 0
+    ): array {
+        // Validate the cmid ID.
+        // TODO: we should maybe pass the bbbid + groupid instead of the cmid.
+        [
+            'cmid' => $cmid,
+            'groupid' => $groupid,
+        ] = self::validate_parameters(self::execute_parameters(), [
+            'cmid' => $cmid,
+            'groupid' => $groupid,
+        ]);
+        $result['warnings'] = [];
+
+        $instance = instance::get_from_cmid($cmid);
+        if (empty($instance)) {
+            throw new \moodle_exception('nosuchinstance', 'mod_bigbluebuttonbn', null,
+                ['entity' => get_string('module', 'course'), 'id' => $cmid]);
+        }
+        if (!groups_group_visible($groupid, $instance->get_course(), $instance->get_cm())) {
+            throw new restricted_context_exception();
+        }
+        $instance->set_group_id($groupid);
+
+        self::validate_context($instance->get_context());
+
+        $returnvalue =
+            meeting::create_and_join_meeting($instance);
+        if (!empty($returnvalue['url']) && empty($returnvalue['warningcode'])) {
+            $result['join_url'] = $returnvalue['url']; // We only return the joinURL if there is no warning or error.
+        }
+        if (!empty($returnvalue['warningcode'])) {
+            $result['warnings'][] = [
+                'item' => 'mod_bigbluebuttonbn',
+                'itemid' => $instance->get_instance_id(),
+                'warningcode' => $returnvalue['warningcode'],
+                'message' => get_string($returnvalue['warningcode'], 'mod_bigbluebuttonbn')
+            ];
+        }
+        if (!empty($returnvalue['errorcode'])) {
```

As noted in the function definition, this should be a try/catch.

---

Not totally sure I understand the comment, but I guess the main issue here is that instead of returning an error we should just raise an exception in the create_and_join_meeting. Is it what you meant ?

---

Yes. We never use the URL if there is an error or a warning, so why have both in the array.